### ISSUE TITLE
[feature] DI in Editors

### ISSUE THREAD

**Issue by jmzagorski** [ghiscoding/aurelia-slickgrid#25]
**I'm submitting a feature request**

* **Browser:**
all
* **Language:**
TypeScript

**Current behavior:**
Library/Custom Editors cannot use Aurelia's DI. External services, such as the `i18n` library, must be injected into the users view model and then added to the `columnDef.params` object just so the grid editor can use that service. Then, the `Editors` have to perform null checks to ensure the user sets the service on the `params` object. This means that our `Editors` using `i18n` (`singleSelectEditor`, `multipleSelectEditor` and `dateTimeEditor`) are dependent on the user passing in the service.

**Expected/desired behavior:**
Library/Custom Editors support Aurelia's DI. To implement this we need to loop through the `columnDefinitions` during the `AureliaCustomElement` binding and reset the `columnDef.editor` property with a wrapper such as:

```typescript
    for (let c of this.columnDefinitions) {
      if (c.editor) {
        c.editor = Factory.of(c.editor).get(this.container)
      }
    }
```

* **What is the motivation / use case for changing the behavior?**
I opened this issue for discussion to get your feedback. I have no need for it right now, but I wonder if it will make for a cleaner API instead of using the `params` property where other values are dumped.

**Pros**
1. Users do not have to inject services into their view model
2 Reduce dependencies for the external service from 2 classes to 1 (2 classes being the Editor and the View Model)

**Cons**
1. We need to take dependencies on the Aurelia's `Container` and `Factory` resolver in the `AureliaCustomElement` (unless you could think of a better way)

**Comment by ghiscoding** [comment]
Some comments and corrections
> External services, such as the i18n library, must be injected into the users view model and then added to the columnDef.params object just so the grid editor can use that service

We can now go simpler since the latest version of SlickGrid, now provide the grid object to both the Formatter and Editor. By doing so, you also have access to the grid options. It's much simpler to pass the `i18n` through the grid options (via `params.i18n`) compare to passing it in every single column def. So the code in the Editor to get grid and grid options is the following (you can see it [here](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/aurelia-slickgrid/src/aurelia-slickgrid/editors/dateEditor.ts#L21) too)
```javascript
init(): void {
  const gridOptions = this.args.grid.getOptions() as GridOption;
}
```

I know it would be easier with DI but Editor are built in property of SlickGrid (which uses it's own Factory). So I'm not sure how to go outside of that, the only possibly way would be to create a new grid option property name like `inlineEditor` that we have full control (as a reference,`filter` was not a built in property of SlickGrid).

So creating a new property, like `inlineEditor` would be possible, you can create your own Factory and DI in it. I'm open to that suggestion

**Comment by jmzagorski** [comment]
Based on my example, SlickGrid can still use its factory and create the editors. All we have to do is use `Factory.of` to wrap their editor class. If we do this we do not have to create any grid options or additional editors. The `Factory` resolver works by passing in all the injected dependencies first (the i18n in this example) and then any additional arguments after, which is the args SlickGrid is passing. I included the diffs below that are required to get this to work and I omitted code so this comment is not so big

**aurelia-slickgrid.ts**
``` diff
- import { Factory, bindable, bindingMode, inject } from 'aurelia-framework';
+ import { Container, Factory, bindable, bindingMode, inject } from 'aurelia-framework';

- @inject(ControlAndPluginService, ExportService, Element, EventAggregator, FilterService, GraphqlService, GridEventService, GridExtraService, I18N, ResizerService, SortService)
+ @inject(Container, ControlAndPluginService, ExportService, Element, EventAggregator, FilterService, GraphqlService, GridEventService, GridExtraService, I18N, ResizerService, SortService)
export class AureliaSlickgridCustomElement {
  
// code omitted for example

  constructor(
+    private container: Container,


  // code omitted for example

  /**
   * Keep original value(s) that could be passed by the user ViewModel.
   * If nothing was passed, it will default to first option of select
   */
  bind(binding: any, contexts: any) {
    // get the grid options (priority is Global Options first, then user option which could overwrite the Global options)
    this.gridOptions = { ...GlobalGridOptions, ...binding.gridOptions };

    this.style = {
      height: `${binding.gridHeight}px`,
      width: `${binding.gridWidth}px`
    };

+    for (let c of this.columnDefinitions) {
+     if (c.editor) {
+      c.editor = Factory.of(c.editor).get(this.container)
+     }
+    }
  }
  // code omitted for example
```

and then 


**singleSelectEditor.ts**
```diff
+@inject(I18N)
export class SingleSelectEditor implements Editor {
  /**
   * The JQuery DOM element
   */
  $editorElm: any;
  /**
   * The slick grid column being edited
   */
  columnDef: Column;
  /**
   * The multiple-select options for a single select
   */
  defaultOptions: any;
  /**
   * The default item value that is set
   */
  defaultValue: any;
  /**
   * The options label/value object to use in the select list
   */
  collection: SelectOption[] = [];
  /**
   * The property name for values in the collection
   */
  valueName: string;
  /**
   * The property name for labels in the collection
   */
  labelName: string;
  /**
   * The i18n aurelia library
   */
  private _i18n: I18N;

-  constructor(private args: any) {
+  constructor(private i18n: I18N, private args: any) {
-     this._i18n = this.args.column.params.i18n;

   // rest of code omitted for example
```

**Comment by ghiscoding** [comment]
That looks interesting, wouldn't mind a PR on that. 
Would you still use the `editor` grid option or a new option name?

**Comment by jmzagorski** [comment]
yes I would use the `editor` property. Nothing _should_ change for the user. But I will work on a PR so we can see it in practice

**Comment by ghiscoding** [comment]
@jmzagorski 
Maybe out of this subject, or not completely, is there any way to build an HTML string from an Aurelia View+ViewModel? I assume that yes through the new SSR (Server Side Rendering). That could be a good thing to look into for the future, just curious if you knew anything about this.

**Comment by jmzagorski** [comment]
I will actually be looking into this more in a about a month or so because my application may require SSR since it will be dealing with the same view models looking differently for the same data.

I am sure there is a better way, but off the top of my head you may be able to use the `ref` binding on the top level element within a `template`. Not sure if you ever used the `ref`, but it is similar to doing `$()` in jquery, but instead with `ref` you get the javascript element. For example

```html
<template>
  <div class="first-parent" ref="$foo">
     <!-- more elements here -->
  </div>
</template>
```

```javascript
export class MyViewModel {
  $foo;

  activate() {
    // give some time for the view to be ready
    setTimeout(() => console.log(this.$foo.outerHTML), 3000)
  }
}
```

**Comment by ghiscoding** [comment]
Ah right `ref`, I was just curious how that would work out with a Formatter that only accept html string.

**Comment by jmzagorski** [comment]
It might be good for us to look at [aurelia-kendoui-bridge](https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge) that I was using. They have a [template](https://github.com/aurelia-ui-toolkits/aurelia-kendoui-bridge/blob/master/src/common/template.js) class that is used in a view and I _think_ their native templates in the actual `kendoui` library use a string too

**Comment by ghiscoding** [comment]
At this point it was just curiosity, but I will add it to the Project tracker without opening an issue. I don't want to open too many issues if we address them in 6 months.

Take a peak at what I'm doing this morning, this makes me smile :)
I just got it working with an input (called it `compoundInputFilter`), now I will have to see if I can combine that with `Flatpickr` for a date filter. I didn't want to look at that now, but our GraphQL implementation throws out error if the date is incorrect (when the user didn't finish typing, dohhh)
![operator](https://user-images.githubusercontent.com/643976/37215572-4e8ffb92-2386-11e8-889b-e54b085f8137.png)

**Comment by jmzagorski** [comment]
👍 that is really nice! I will be using that in my app when it is finished.

**Comment by ghiscoding** [comment]
@jmzagorski 
So I made a proof of concept today for the Grouping & Aggregators, in about 4 hours (this morning), in my Angular repo which you can see a demo [here](https://ghiscoding.github.io/Angular-Slickgrid/#/grouping). It's really quick and dirty, which is why I tagged it as Alpha stage, I'll work it out in the weekend for Aurelia.

It looks like I am facing again the same issue with DI since I need the i18n yet again with the Aggregators. I assume this will be similar to this issue. So we might have to re-use your DI implementation for other areas, whenever that is ready. 

Really no rush, I'm just writing here as reference before I forget. I have already opened an issue in SlickGrid to expose the grid object to these GroupTotalsFormatter so we'll see how it goes.

**Comment by jmzagorski** [comment]
this should be easy to implement. I will plan to submit a PR tomorrow.

**Comment by ghiscoding** [comment]
Closed by PR #33
